packages/gatsby-recipes/src/providers/gatsby/sitemetadata.js

```diff
@@ -0,0 +1,274 @@
+    key: name,
+    value: undefined,
+  })
+
+  const { code } = babel.transform(src, {
+    plugins: [addPlugins.plugin],
+    configFile: false,
+  })
+
+  return code
+}
+
+const getSiteMetdataFromConfig = src => {
+  const getSiteMetadata = new BabelPluginGetSiteMetadataFromConfig()
+
+  babel.transform(src, {
+    plugins: [getSiteMetadata.plugin],
+    configFile: false,
+  })
+
+  return getSiteMetadata.state
+}
+
+const getConfigPath = root => path.join(root, `gatsby-config.js`)
+
+const readConfigFile = async root => {
+  let src
+  try {
+    src = await fs.readFile(getConfigPath(root), `utf8`)
+  } catch (e) {
+    if (e.code === `ENOENT`) {
+      src = `/**
+ * Configure your Gatsby site with this file.
+ *
+ * See: https://www.gatsbyjs.org/docs/gatsby-config/
+ */
+
+module.exports = {
+  plugins: [],
+}`
+    } else {
+      throw e
+    }
+  }
+
+  return src
+}
+
+const create = async ({ root }, { name, value }) => {
+  const configSrc = await readConfigFile(root)
+  const prettierConfig = await prettier.resolveConfig(root)
+
+  let code = addFieldToSiteMetadata(configSrc, { name, value })
+  code = prettier.format(code, { ...prettierConfig, parser: `babel` })
+
+  await fs.writeFile(getConfigPath(root), code)
+
+  return await read({ root }, name)
+}
+
+const read = async ({ root }, id) => {
+  try {
+    const configSrc = await readConfigFile(root)
+
+    const siteMetadata = getSiteMetdataFromConfig(configSrc)
+
+    if (!siteMetadata || typeof siteMetadata[id] === undefined) return null
+
+    return {
+      name: id,
+      value: JSON.stringify(siteMetadata[id]),
+    }
+  } catch (e) {
+    console.log(e)
+    throw e
+  }
+}
+
+const destroy = async ({ root }, resource) => {
+  const configSrc = await readConfigFile(root)
+
+  const newSrc = removeFieldFromSiteMetadata(configSrc, resource)
+
+  await fs.writeFile(getConfigPath(root), newSrc)
+}
+
+class BabelPluginSetSiteMetadataField {
+  constructor({ key, value }) {
+    this.plugin = declare(api => {
+      api.assertVersion(7)
+
+      return {
+        visitor: {
+          ExpressionStatement(path) {
+            const { node } = path
+            const { left, right } = node.expression
+
+            if (!isDefaultExport(left)) {
+              return
+            }
+
+            const siteMetadata = right.properties.find(
+              p => p.key.name === `siteMetadata`
+            )
+
+            if (!siteMetadata || !siteMetadata.value) return
+
+            const siteMetadataObj = getObjectFromNode(siteMetadata.value)
+
+            const newSiteMetadataObj = {
+              ...siteMetadataObj,
+              [key]: typeof value === `string` ? JSON.parse(value) : value,
+            }
+
+            const newSiteMetadataTemplate = template(`
+              const foo = ${JSON.stringify(newSiteMetadataObj, null, 2)}
+            `)()
+
+            const newSiteMetadata = newSiteMetadataTemplate.declarations[0].init
```

Comment:
Should this use optional chaining? That's a lot of assumed existence.

Reply:
We control the code that is parsed (the `const foo` thing), so we know this will always exist! 👍